src/coreclr/tools/r2rdump/CoreDisTools.cs

```diff
@@ -1209,6 +1213,235 @@ private static bool IsAnotherRuntimeFunctionWithinMethod(int rva, RuntimeFunctio
             return false;
         }
 
+        /// <summary>
+        /// Improves disassembler output for RiscV64 by adding comments at the end of instructions.
+        /// </summary>
+        /// <param name="rtf">Runtime function</param>
+        /// <param name="imageOffset">Offset within the image byte array</param>
+        /// <param name="rtfOffset">Offset within the runtime function</param>
+        /// <param name="instruction">Textual representation of the instruction</param>
+        private void ProbeRiscV64Quirks(RuntimeFunction rtf, int imageOffset, int rtfOffset, ref string instruction)
+        {
+            const int InstructionSize = 4;
+            uint instr = BitConverter.ToUInt32(_reader.Image, imageOffset + rtfOffset);
+
+            /*
+                Pattern:
+                    auipc
+                    addi
+                    ld
+                    jalr
+                , where addi and ld may be repeated many times.
+                Instructions that have no effect on the jump 
+                address calculation are skipped.
+            */
+            if (IsRiscV64JalrInstruction(instr))
+            {
+                AnalyzeRiscV64Itype(instr, out uint rd, out uint rs1, out int imm);
+                uint register = rs1;
+                int immValue = imm;
+
+                int target = 0;
+                bool isFound = false;
+                int currentInstrOffset = rtfOffset - InstructionSize;
+                // skip all subsequent ld,lw instructions until first non-ld,lw instruction
+                do
+                {
+                    instr = BitConverter.ToUInt32(_reader.Image, imageOffset + currentInstrOffset);
+
+                    if (IsRiscV64LdInstruction(instr) || IsRiscV64LwInstruction(instr))
```

Comment:
I am just curious. In your examples, check for `lw` instruction is not necessary. Could you give an example with `lw` instruction? Thank you.

Reply:
As i know there is no function call which uses `lw` to calculate jump address. However, this check for `lw` is necessary because this first `if` is trying to skip all `ld` and `lw` instructions which are used to push arguments (likely). For example
```
int Interop+Globalization.GetCalendars(string, System.Globalization.CalendarId[], int)
...
14935c: 00e22517    auipc   a0, 3618
149360: c5450513    addi    a0, a0, -940
149364: 00053503    ld      a0, 0(a0)
149368: 00053683    ld      a3, 0(a0)
14936c: fe443503    ld      a0, -28(s0)
149370: fec42603    lw      a2, -20(s0)
149374: fdc43583    ld      a1, -36(s0)
149378: fd443f03    ld      t5, -44(s0)
14937c: 000680e7    jalr    a3 // <calle name>
....

bool Interop+Globalization.EnumCalendarInfo(IntPtr, string, System.Globalization.CalendarId, System.Globalization.CalendarDataType, IntPtr)n
...
149470: 00e22517    auipc   a0, 3618
149474: b5050513    addi    a0, a0, -1200
149478: 00053503    ld      a0, 0(a0)
14947c: 00053783    ld      a5, 0(a0)
149480: fe843583    ld      a1, -24(s0)
149484: ff442603    lw      a2, -12(s0)
149488: fe043503    ld      a0, -32(s0)
14948c: fd843703    ld      a4, -40(s0)
149490: ff042683    lw      a3, -16(s0)
149494: fd043f03    ld      t5, -48(s0)
149498: 000780e7    jalr    a5 // <calle name>
...

void Interop+Globalization.InitOrdinalCasingPage(int, char*)
...
149634: 00e22517    auipc   a0, 3618
149638: 91c50513    addi    a0, a0, -1764
14963c: 00053503    ld      a0, 0(a0)
149640: 00053603    ld      a2, 0(a0)
149644: ff443583    ld      a1, -12(s0)
149648: ffc42503    lw      a0, -4(s0)
14964c: fec43f03    ld      t5, -20(s0)
149650: 000600e7    jalr    a2 // <calle name>
...

int Interop+Globalization.CompareString(IntPtr, char*, int, char*, int, System.Globalization.CompareOptions)
...
149aa4: 00e21517    auipc   a0, 3617
149aa8: 4bc50513    addi    a0, a0, 1212
149aac: 00053503    ld      a0, 0(a0)
149ab0: 00053803    ld      a6, 0(a0)
149ab4: ffc42783    lw      a5, -4(s0)
149ab8: fec43503    ld      a0, -20(s0)
149abc: fe443583    ld      a1, -28(s0)
149ac0: ff842603    lw      a2, -8(s0)
149ac4: fdc43683    ld      a3, -36(s0)
149ac8: ff442703    lw      a4, -12(s0)
149acc: fd443f03    ld      t5, -44(s0)
149ad0: 000800e7    jalr    a6 // <calle name>
``` 
As you can see, the sequence of `lw` and `ld` instructions is not fixed.

Reply:
Well. IMO, `lw` instructions can be skipped in line 1334 in `StaticAnalyzeRiscV64Assembly`. Am I wrong?
```
else
{
    // check if "register" is calculated using an unsupported instruction
    uint rd = (instr >> 7) & 0b_11111U;
    if (rd == register)
    {
        return false;
    }
}
```

Reply:
if we want to skip `lw` in `StaticAnalyzeRiscV64Assembly`, we will have to skip `ld` too. In other case, we would have a problem, for example, with this:
```
void Interop+Globalization.InitOrdinalCasingPage(int, char*)
...
149634: 00e22517    auipc   a0, 3618
149638: 91c50513    addi    a0, a0, -1764
14963c: 00053503    ld      a0, 0(a0)
149640: 00053603    ld      a2, 0(a0)
149644: ff443583    ld      a1, -12(s0)
149648: ffc42503    lw      a0, -4(s0)
14964c: fec43f03    ld      t5, -20(s0)
149650: 000600e7    jalr    a2 // <calle name>
```

Reply:
I see. You are right. However, actually what I want is removing 'lw' and implementing simpler like #102146. Could you check loongarch PR and share your idea? Thank you.

Reply:
I simplify code to calculate callee address

Reply:
Thank you

Reply:
no problem :D